### ISSUE TITLE
Disallow additional properties?

### ISSUE THREAD

**Issue by iSplasher** [sinclairzx81/typebox#56]
Hello, do you have any idea as to how I would disallow additional properties when validating?
I've read through ajv and found `additionalProperties: false`.
```
function TObject<T extends TProperties>(props: T, options?: CustomOptions) {
  return Type.Object<T>(props, { ...options, additionalProperties: false });
}
```
I've built my schema up to look like this:
`const AttrObject = Type.Dict(AttributeValue);`
Where `AttributeValue` is built up to look like this:
![billede](https://user-images.githubusercontent.com/9113348/111084595-6f1d5c00-8513-11eb-899a-09570fa91b04.png)
Ignore the "value" key.

Validating against this should work, but I still get errors?
```json
{
   "field1": { "type": "string"  }
}
```

Here's the raw json of the schema: https://pastebin.com/E1xwx1C5
And here are the errors I get: https://pastebin.com/iEYsekBH

**Comment by iSplasher** [comment]
I've found a workaround.
If I "flatten" the type then it'll work as expected.
Flattening means that I take intersects and unions and merge them into a single flat object type.
Again, this code is very specific to my use case but here's an example of the flattening utility:
```ts
function flattenObject<T extends ObjectType>(value: T, _s = -1) {
  // eslint-disable-next-line no-param-reassign
  _s += 1;

  let obj: any = {};

  if (value.kind === ObjectKind) {
    Object.assign(
      obj,
      ...Object.entries(value.properties).map(([k, v]: any[]) => ({
        [k]: [ObjectKind, UnionKind, IntersectKind, DictKind].includes(v)
          ? flattenObject(v, _s)
          : v,
      }))
    );

    if (!_s) {
      obj = TObject(obj);
    }

    if (value.required) {
      for (const k of value.required) {
        if (!value.properties[k]) {
          obj.required = [...(obj.required ?? []), k];
        }
      }
    }
  }

  // must be intersect
  if (value.allOf) {
    Object.assign(obj, ...value.allOf.map((i: any) => flattenObject(i, _s)));
  }

  // must be union
  if (value.anyOf) {
    Object.assign(obj, ...value.anyOf.map((i: any) => flattenObject(i, _s)));
  }

  if (!_s && obj.kind !== ObjectKind) {
    obj = TObject(obj);
  }

  const removeRequired = (o: ReturnType<typeof Type.Object>) => {
    if (o?.properties?.required) {
      o.required = [...(o.required ?? []), ...(o?.properties?.required as any)];
      delete o?.properties.required;
    }

    if (o?.required) {
      o.required = o.required.filter((k) => k !== 'required');
    }
  };

  removeRequired(obj);

  return obj as typeof AttributeValueT;
}
```

This is the TS utility to infer the correct type with Static types:
```
// Flatten an intersection type by merging keys:
export type Flatten<T> = T extends Record<string, any>
  ? { [k in keyof T]: T[k] }
  : never;
```

You may close the issue

**Comment by sinclairzx81** [comment]
@iSplasher Hi, thanks that's all good. Just had a chance to look at your issue. Now you mention it the `TObject<TProperties>` should probably disallow `additionalProperties` by default inline with TypeScript. For instance...

```typescript
interface Foo { x: number, y: number }

function test(foo: Foo) {}

test({ x: 1, y: 2, z: 3 }) 

// ^ Argument of type '{ x: number; y: number; z: number; }' is 
// not assignable to parameter of type 'Foo'.
``` 

Looking at `Type.Object(...)` It is not applying `additionalProperties`

```typescript
/** Creates a `object` schema with the given properties. */
public Object<T extends TProperties>(properties: T, options: CustomOptions = {}): TObject<T> {
	const property_names = Object.keys(properties)
	const optional = property_names.filter(name => {
		const candidate = properties[name] as TModifier
		return (candidate.modifier &&
			(candidate.modifier === OptionalModifier ||
				candidate.modifier === ReadonlyOptionalModifier))
	})
	const required_names = property_names.filter(name => !optional.includes(name))
	const required = (required_names.length > 0) ? required_names : undefined
	return (required) ? 
		{ ...options, kind: ObjectKind, type: 'object', properties, required } : 
		{ ...options, kind: ObjectKind, type: 'object', properties }
	//    ^ additionalProperties are missing here. Can be added on CustomOptions
}
```
Could consider this a bug / omission. I expect there would be less hassle for users who've had to add `additionalProperties: false` and I expect having `additionalProperties: false` is more the rule than the exception.

Will leave this issue open for a while and have a think about it. Making this change may be a breaking change for some users.

Thanks for reporting!
S

**Comment by sinclairzx81** [comment]
@iSplasher Have submitted a AJV bug report around this https://github.com/ajv-validator/ajv/issues/1496 . There may be an issue with default enabling `additionalProperties: false` for `Type.Object(...)` as it breaks validation for `Type.Intersect(...)`. There might be a bit of ambiguity in the JSON schema spec around this. Will see what the AJV contributors come back with.

**Comment by sinclairzx81** [comment]
@iSplasher Have made the updates to resolve this issue. `Type.Object(...)` now has `additionalProperties: false` by default. This update also includes implementation / schema changes to `Type.Intersect(...)` based on feedback from the `AJV` contributors. Updates are published to `@sinclair/typebox@0.15.0`.

Will close off this issue :) 
Cheers mate
S
